concurrent_map.go

```diff
@@ -5,75 +5,93 @@ import (
 	"sync"
 )
 
-var SHARD_COUNT = 32
+// ShardCount represent the number of shard the map is divided into.
+var ShardCount = 32
 
-// A "thread" safe map of type string:Anything.
-// To avoid lock bottlenecks this map is dived to several (SHARD_COUNT) map shards.
+// ConcurrentMap is a "thread" safe map of type string:Anything.
+// To avoid lock bottlenecks this map is dived to several (ShardCount) map shards.
 type ConcurrentMap []*ConcurrentMapShared
 
-// A "thread" safe string to anything map.
+// ConcurrentMapShared is a "thread" safe string to anything map.
 type ConcurrentMapShared struct {
+	mapKeys      []string
+	maxSize      int
 	items        map[string]interface{}
 	sync.RWMutex // Read Write mutex, guards access to internal map.
 }
 
-// Creates a new concurrent map.
-func New() ConcurrentMap {
-	m := make(ConcurrentMap, SHARD_COUNT)
-	for i := 0; i < SHARD_COUNT; i++ {
-		m[i] = &ConcurrentMapShared{items: make(map[string]interface{})}
+// New creates a new concurrent map.
+func New(maxSize int) ConcurrentMap {
+	m := make(ConcurrentMap, ShardCount)
+	shardKeySize := maxSize / ShardCount
+	for i := 0; i < ShardCount; i++ {
+		m[i] = &ConcurrentMapShared{
+			maxSize: shardKeySize,
+			items:   make(map[string]interface{}),
+		}
 	}
 	return m
 }
 
-// GetShard returns shard under given key
+// GetShard returns shard under given key.
 func (m ConcurrentMap) GetShard(key string) *ConcurrentMapShared {
-	return m[uint(fnv32(key))%uint(SHARD_COUNT)]
+	return m[uint(fnv32(key))%uint(ShardCount)]
 }
 
 func (m ConcurrentMap) MSet(data map[string]interface{}) {
 	for key, value := range data {
-		shard := m.GetShard(key)
-		shard.Lock()
-		shard.items[key] = value
-		shard.Unlock()
+		m.Set(key, value)
 	}
 }
 
-// Sets the given value under the specified key.
+// Set sets the given value under the specified key.
 func (m ConcurrentMap) Set(key string, value interface{}) {
 	// Get map shard.
 	shard := m.GetShard(key)
 	shard.Lock()
+	if m.isEvictionOccurred(shard) {
```

Comment:
Same problem with eviction (which might not be required). If the key is already added, setting the new value might result in eviction and re-appending the same key (which might result in double keys problem)

Reply:
I have completely removed it, since we are checking for eviction on ccache `Put` method.

Reply:
Not ok as Set and SetIfAbsent are exported. Calling those methods will create inconsistencies in the code. Either remove the exported funcs or revert the eviction **but** the eviction should be done only if it needs to add the new value.